src/runtime/virtcontainers/clh.go

```diff
@@ -882,16 +966,64 @@ func (clh *cloudHypervisor) Cleanup(ctx context.Context) error {
 
 func (clh *cloudHypervisor) PauseVM(ctx context.Context) error {
 	clh.Logger().WithField("function", "PauseVM").Info("Pause Sandbox")
+
+	cl := clh.client()
+
+	_, err := cl.PauseVM(ctx)
+	if err != nil {
+		return openAPIClientError(err)
+	}
+
 	return nil
 }
 
 func (clh *cloudHypervisor) SaveVM() error {
 	clh.Logger().WithField("function", "saveSandboxC").Info("Save Sandbox")
+
+	cl := clh.client()
+
+	_, err := clh.vmInfo()
+	if err != nil {
+		clh.Logger().WithField("function", "SaveVM").WithError(err).Info("[clh] vmInfo failed")
+		return openAPIClientError(err)
+	}
+
+	snaopshotConfig := *chclient.NewVmSnapshotConfig()
+	destinationUrl := "file://" + filepath.Dir(clh.config.DevicesStatePath)
+	snaopshotConfig.DestinationUrl = &destinationUrl
+	_, err = cl.VmSnapshotPut(clh.ctx, snaopshotConfig)
+	if err != nil {
+		return openAPIClientError(err)
+	}
+	clh.Logger().WithField("function", "saveSandboxC").Info("Save Sandbox finish")
 	return nil
 }
 
 func (clh *cloudHypervisor) ResumeVM(ctx context.Context) error {
 	clh.Logger().WithField("function", "ResumeVM").Info("Resume Sandbox")
+
+	cl := clh.client()
+
+	_, err := cl.ResumeVM(ctx)
+	if err != nil {
+		return openAPIClientError(err)
+	}
+
+	return nil
+}
+
+func (clh *cloudHypervisor) restoreFromTemplate(statePath string) error {
```

Comment:
- What if `statePath == ""`?

Reply:
Also, does it need to be absolute?

Reply:
Thanks, already add statePath == "" and absolute path check.